File: C#/Hashmap & Set/1657. Determine if Two Strings Are Close.cs
/* 1657. Determine if Two Strings Are Close
Difficulty: Medium
Submission Link: https://leetcode.com/problems/determine-if-two-strings-are-close/submissions/1146084467/ */

public class Solution {
    public bool CloseStrings(string word1, string word2) {
        int l1 = word1.Length, l2 = word2.Length;
        if (l1 != l2) return false;
        if (word1.Equals(word2)) return true;

        int[] firstSet = new int[26];
        int[] secondSet = new int[26];
        
        for (int i = 0; i < l1; i++) {
            firstSet[word1[i] - 'a']++;
        }
        
        for (int i = 0; i < l2; i++) {
            if (firstSet[word2[i] - 'a'] == 0) return false;
            secondSet[word2[i] - 'a']++;
        }
        
        Array.Sort(firstSet);
        Array.Sort(secondSet);

        for (int i = 25; i >= 0 && firstSet[i] != 0; i--) {
            if (firstSet[i] != secondSet[i]) {
                return false;
            }
        }
        
        return true;
    }
}

File: Python3/Hashmap & Set/1657. Determine if Two Strings Are Close.py
""" 1657. Determine if Two Strings Are Close
Difficulty: Medium
Submission Link: https://leetcode.com/problems/determine-if-two-strings-are-close/submissions/1146085063/ """

class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        # Step 1: Check if lengths are equal
        if len(word1) != len(word2):
            return False

        # Step 2: Check if strings are already equal
        if word1 == word2:
            return True

        # Step 3: Count occurrences of each character
        first_set = [0] * 26
        second_set = [0] * 26
        
        for ch in word1:
            first_set[ord(ch) - ord('a')] += 1

        for ch in word2:
            if first_set[ord(ch) - ord('a')] == 0:
                return False
            second_set[ord(ch) - ord('a')] += 1

        # Step 4: Check if sets of characters are the same
        if set(first_set) != set(second_set):
            return False

        # Step 5: Sort both arrays
        first_set.sort()
        second_set.sort()

        # Step 6: Compare sorted arrays element by element
        for i in range(26):
            if first_set[i] != second_set[i]:
                return False

        # Step 7: If all conditions are satisfied, return True
        return True